rpcs3/Emu/Cell/Modules/cellMic.cpp

```diff
@@ -3,180 +3,380 @@
+	return micThread->micOpened;
 }
 
-s32 cellMicIsAttached()
+s32 cellMicIsAttached(u32 deviceNumber)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
-	return CELL_OK;
+	cellMic.warning("cellMicIsAttached(deviceNumber=%d)", deviceNumber);
+	return 1;
 }
 
-s32 cellMicIsOpen()
+s32 cellMicClose(u32 deviceNumber)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicClose(deviceNumber=%d)", deviceNumber);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
+		return CELL_MIC_ERROR_NOT_INIT;
+
+	micThread->micOpened = false;
 	return CELL_OK;
 }
 
-s32 cellMicGetDeviceAttr()
+/// Starting/Stopping Microphone Functions
+
+s32 cellMicStart(u32 deviceNumber)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicStart(deviceNumber=%d)", deviceNumber);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
+		return CELL_MIC_ERROR_NOT_INIT;
+	if (!micThread->micOpened)
+		return CELL_MIC_ERROR_NOT_OPEN;
+
+	micThread->micStarted = true;
 	return CELL_OK;
 }
 
-s32 cellMicSetDeviceAttr()
+s32 cellMicStartEx(u32 deviceNumber, u32 flags)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicStartEx(deviceNumber=%d)", deviceNumber);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
+		return CELL_MIC_ERROR_NOT_INIT;
+	if (!micThread->micOpened)
+		return CELL_MIC_ERROR_NOT_OPEN;
+
+	micThread->micStarted = true;
 	return CELL_OK;
 }
 
-s32 cellMicGetSignalAttr()
+s32 cellMicStop(u32 deviceNumber)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicStop(deviceNumber=%d)", deviceNumber);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
+		return CELL_MIC_ERROR_NOT_INIT;
+	if (!micThread->micOpened)
+		return CELL_MIC_ERROR_NOT_OPEN;
+	micThread->micStarted = false;
 	return CELL_OK;
 }
 
-s32 cellMicSetSignalAttr()
+/// Microphone Attributes/States Functions
+
+s32 cellMicGetDeviceAttr(u32 deviceNumber, vm::ptr<void> deviceAttributes, vm::ptr<u32> arg1, vm::ptr<u32> arg2)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicGetDeviceAttr(deviceNumber=%d, deviceAttribute=0x%x, arg1=0x%x, arg2=0x%x)", deviceNumber, deviceAttributes, arg1, arg2);
 	return CELL_OK;
 }
 
-s32 cellMicGetSignalState()
+s32 cellMicSetDeviceAttr()
 {
 	UNIMPLEMENTED_FUNC(cellMic);
 	return CELL_OK;
 }
 
-s32 cellMicStart()
+s32 cellMicGetSignalAttr()
 {
 	UNIMPLEMENTED_FUNC(cellMic);
 	return CELL_OK;
 }
 
-s32 cellMicRead()
+s32 cellMicSetSignalAttr()
 {
 	UNIMPLEMENTED_FUNC(cellMic);
 	return CELL_OK;
 }
 
-s32 cellMicStop()
+s32 cellMicGetSignalState(u32 deviceNumber, CellMicSignalState signalState, vm::ptr<void> value)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicGetSignalState(deviceNumber=%d, signalSate=%d, value=0x%x)", deviceNumber, (int) signalState,
+				 value);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
+		return CELL_MIC_ERROR_NOT_INIT;
+
+	be_t<u32> *ival = (be_t<u32> *) value.get_ptr();
+	f32 *fval = (f32 *) value.get_ptr();
```

Comment:
why aren't you using be_t for the float here?

Reply:
I haven't tested to make sure that it should be be_t

Reply:
seems likely that it should be be_t, youre reading from ps3 memory after all